flink-runtime/src/main/java/org/apache/flink/runtime/shuffle/NettyShuffleUtils.java

```diff
@@ -0,0 +1,138 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.shuffle;
+
+import org.apache.flink.runtime.io.network.partition.ResultPartitionType;
+import org.apache.flink.runtime.jobgraph.IntermediateDataSetID;
+
+import org.apache.commons.lang3.tuple.Pair;
+
+import java.util.Map;
+
+import static org.apache.flink.util.Preconditions.checkArgument;
+
+/**
+ * Utils to calculate network memory requirement of a vertex from network configuration and details
+ * of input and output. The methods help to decide the volume of buffer pools when initializing
+ * shuffle environment and also guide network memory announcing in fine-grained resource management.
+ */
+public class NettyShuffleUtils {
+
+    public static Pair<Integer, Integer> getMinMaxFloatingBuffersPerInputGate(
+            final int numFloatingBuffersPerGate) {
+        // We should guarantee at-least one floating buffer for local channel state recovery.
+        return Pair.of(1, numFloatingBuffersPerGate);
+    }
+
+    public static Pair<Integer, Integer> getMinMaxNetworkBuffersPerResultPartition(
+            final int numBuffersPerChannel,
+            final int numFloatingBuffersPerGate,
+            final int sortShuffleMinParallelism,
+            final int sortShuffleMinBuffers,
+            final int numSubpartitions,
+            final ResultPartitionType type) {
+        int min =
+                type.isBlocking() && numSubpartitions >= sortShuffleMinParallelism
+                        ? sortShuffleMinBuffers
+                        : numSubpartitions + 1;
+        int max =
+                type.isBounded()
+                        ? numSubpartitions * numBuffersPerChannel + numFloatingBuffersPerGate
+                        : Integer.MAX_VALUE;
+        return Pair.of(min, max);
+    }
+
+    public static int computeNetworkBuffersForAnnouncing(
+            final int numBuffersPerChannel,
+            final int numFloatingBuffersPerGate,
+            final int sortShuffleMinParallelism,
+            final int sortShuffleMinBuffers,
+            final int numTotalInputChannels,
+            final int numTotalInputGates,
+            final Map<IntermediateDataSetID, Integer> subpartitionNums,
+            final Map<IntermediateDataSetID, ResultPartitionType> partitionTypes) {
+
+        // Each input channel will retain N exclusive network buffers, N = numBuffersPerChannel.
+        // Each input gate is guaranteed to have a number of floating buffers.
+        int requirmentForInputs =
+                numBuffersPerChannel * numTotalInputChannels
+                        + numFloatingBuffersPerGate * numTotalInputGates;
```

Comment:
To keep consistency, I think `getMinMaxFloatingBuffersPerInputGate` should be changed to `getMinMaxFloatingBuffersPerInputGate (getMinMaxFloatingBuffersPerInputGate).getRight()`

Reply:
That's better, thanks ~